src/chunk_data_node.c

```diff
@@ -0,0 +1,311 @@
+/*
+ * This file and its contents are licensed under the Apache License 2.0.
+ * Please see the included NOTICE for copyright information and
+ * LICENSE-APACHE for a copy of the license.
+ */
+#include <postgres.h>
+#include <catalog/pg_foreign_table.h>
+#include <catalog/pg_foreign_server.h>
+#include <catalog/dependency.h>
+#include <foreign/foreign.h>
+#include <utils/builtins.h>
+#include <utils/syscache.h>
+#include <utils/inval.h>
+#include <access/xact.h>
+
+#include "chunk_data_node.h"
+#include "scanner.h"
+#include "chunk.h"
+
+static void
+chunk_data_node_insert_relation(Relation rel, int32 chunk_id, int32 node_chunk_id, Name node_name)
+{
+	TupleDesc desc = RelationGetDescr(rel);
+	Datum values[Natts_chunk_data_node];
+	bool nulls[Natts_chunk_data_node] = { false };
+	CatalogSecurityContext sec_ctx;
+
+	values[AttrNumberGetAttrOffset(Anum_chunk_data_node_chunk_id)] = Int32GetDatum(chunk_id);
+	values[AttrNumberGetAttrOffset(Anum_chunk_data_node_node_chunk_id)] =
+		Int32GetDatum(node_chunk_id);
+	values[AttrNumberGetAttrOffset(Anum_chunk_data_node_node_name)] = NameGetDatum(node_name);
+
+	ts_catalog_database_info_become_owner(ts_catalog_database_info_get(), &sec_ctx);
+	ts_catalog_insert_values(rel, desc, values, nulls);
+	ts_catalog_restore_user(&sec_ctx);
+}
+
+static void
+chunk_data_node_insert_internal(int32 chunk_id, int32 node_chunk_id, Name node_name)
+{
+	Catalog *catalog = ts_catalog_get();
+	Relation rel;
+
+	rel = heap_open(catalog->tables[CHUNK_DATA_NODE].id, RowExclusiveLock);
+
+	chunk_data_node_insert_relation(rel, chunk_id, node_chunk_id, node_name);
+
+	heap_close(rel, RowExclusiveLock);
+}
+
+void
+ts_chunk_data_node_insert(ChunkDataNode *node)
+{
+	chunk_data_node_insert_internal(node->fd.chunk_id, node->fd.node_chunk_id, &node->fd.node_name);
+}
+
+void
+ts_chunk_data_node_insert_multi(List *chunk_data_nodes)
+{
+	Catalog *catalog = ts_catalog_get();
+	Relation rel;
+	ListCell *lc;
+
+	rel = heap_open(catalog->tables[CHUNK_DATA_NODE].id, RowExclusiveLock);
```

Comment:
Should it be `relation_open` as `heap_open` was already replaced in singlenode code?
```suggestion
	rel = relation_open(catalog->tables[CHUNK_DATA_NODE].id, RowExclusiveLock);
```